lib/bundler/definition.rb

```diff
@@ -581,6 +586,7 @@ def dependencies_for_source_changed?(source, locked_source = source)
     end
 
     def specs_for_source_changed?(source)
+      return false if source.is_a?(Source::Path) && !source.has_spec_files? && Bundler.settings[:frozen]
```

Comment:
I think this was trying to address a different issue, and so it needs test coverage